File: database-driver-plugin/src/main/java/de/lino/database/utility/cache/DefaultCache.java
package de.lino.database.utility.cache;

import de.lino.database.utils.cache.Cache;

import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Function;

/**
 * Default, in-memory implementation of {@link Cache}, backed by a
 * {@link ConcurrentHashMap}.
 * <p>
 * <b>Complexity:</b>
 * <ul>
 *   <li>{@link #get}: O(1) amortized (hash lookup in {@link ConcurrentHashMap}).
 *       Worst case O(n) only under pathological hash collisions — practically irrelevant.</li>
 *   <li>{@link #put} / {@link #invalidate}: O(1) amortized.</li>
 *   <li>{@link #evictExpired()}: O(n) — a full scan, so it should NOT be called on the
 *       hot path but periodically (e.g. every 60s via a scheduler).</li>
 *   <li>Size-based eviction (when {@code maxSize} is set): O(k) with a constant
 *       sample factor k (default 8), <b>independent of n</b> — no O(n) scan on every
 *       insert. This mirrors Redis' "approximated LRU": instead of locating the exact
 *       oldest entry (O(n), or O(log n) with an extra data structure), a small random
 *       sample is compared and the oldest candidate within it is evicted. That keeps
 *       the insert path constant-time even with very many entries.</li>
 * </ul>
 * <p>
 * <b>Scaling (clustering / volume):</b>
 * {@link ConcurrentHashMap} itself is unproblematic up to roughly 10<sup>7</sup>–10<sup>8</sup>
 * entries (segmented locking, no global lock on reads/writes). In practice the limiting
 * factor is the heap, not the algorithm:
 * <ul>
 *   <li>~10<sup>4</sup>–10<sup>6</sup> entries: fine with a default JVM heap, no {@code maxSize} needed.</li>
 *   <li>~10<sup>6</sup>–10<sup>7</sup> entries: use {@code maxSize} plus sample-based eviction to bound
 *       growth; a rough overhead estimate is 100–150 bytes per entry on top of the entity
 *       itself (map node + entry wrapper + {@link CompletableFuture}) — so ~1–1.5 GB of pure
 *       overhead at 10<sup>7</sup> entries.</li>
 *   <li>&gt;10<sup>8</sup> entries, or the need for a cache shared across multiple JVM instances
 *       ("real" clustering): this in-memory cache is not designed for that — an external,
 *       distributed cache (e.g. Redis, Hazelcast, Infinispan) is recommended instead, since a
 *       single JVM heap instance would otherwise suffer from GC pauses.</li>
 * </ul>
 * <p>
 * <b>Further design decisions:</b>
 * <ul>
 *   <li><b>Stampede protection:</b> {@link ConcurrentHashMap#compute} ensures that under
 *       concurrent requests for the same key, the loader runs only ONCE — all callers
 *       share the same future.</li>
 *   <li><b>Async:</b> {@link #get} never blocks; the loader supplies a
 *       {@link CompletableFuture} that the caller can chain freely.</li>
 *   <li><b>Usability:</b> a single read method ({@code get}); TTL and {@code maxSize}
 *       are optional and configurable per cache instance.</li>
 *   <li><b>Null-safety:</b> required parameters are checked via
 *       {@link Objects#requireNonNull} at the top of each method.</li>
 * </ul>
 *
 * @param <ID> type of the key (e.g. Long, UUID, String)
 * @param <T>  type of the cached entity
 */
public final class DefaultCache<ID, T> implements Cache<ID, T> {

    private static final int EVICTION_SAMPLE_SIZE = 8;

    private final ConcurrentHashMap<ID, CompletableFuture<Entry<T>>> store = new ConcurrentHashMap<>();
    private final Function<ID, CompletableFuture<T>> loader;
    private final Duration ttl;
    private final long maxSize; // <= 0 means unbounded

    private static final class Entry<T> {

        final T value;
        final Instant expiresAt; // null = never expires
        final AtomicLong lastAccessNanos = new AtomicLong(System.nanoTime());

        Entry(T value, Instant expiresAt) {
            this.value = value;
            this.expiresAt = expiresAt;
        }

        boolean isExpired() {
            return expiresAt != null && Instant.now().isAfter(expiresAt);
        }

        void touch() {
            lastAccessNanos.set(System.nanoTime());
        }

    }

    /**
     * @param loader  asynchronously supplies an entity when it is not (or no longer) cached
     * @param ttl     validity duration of an entry, {@code null} for unbounded
     * @param maxSize maximum number of entries; {@code <= 0} for unbounded.
     *                Approximate-LRU eviction (O(k)) is triggered once this is exceeded.
     */
    public DefaultCache(Function<ID, CompletableFuture<T>> loader, Duration ttl, long maxSize) {
        this.loader = Objects.requireNonNull(loader, "loader must not be null");
        this.ttl = ttl;
        this.maxSize = maxSize;
    }

    /**
     * @param loader asynchronously supplies an entity when it is not (or no longer) cached
     * @param ttl    validity duration of an entry, {@code null} for unbounded
     */
    public DefaultCache(Function<ID, CompletableFuture<T>> loader, Duration ttl) {
        this(loader, ttl, -1);
    }

    @Override
    public CompletableFuture<T> get(ID id) {
        Objects.requireNonNull(id, "id must not be null");

        CompletableFuture<Entry<T>> future = store.compute(id, (key, existing) -> {
            if (existing != null && !isFailedOrExpired(existing)) {
                return existing;
            }
            return loader.apply(key).thenApply(value -> {
                Objects.requireNonNull(value, "loader must not return null values");
                return new Entry<>(value, ttl == null ? null : Instant.now().plus(ttl));
            });
        });

        maybeEvictOverCapacity();

        return future.thenApply(entry -> {
            entry.touch(); // update access time for approximate LRU
            return entry.value;
        });
    }

    private boolean isFailedOrExpired(CompletableFuture<Entry<T>> future) {
        if (!future.isDone()) {
            return false; // still loading — piggyback instead of restarting
        }
        if (future.isCompletedExceptionally()) {
            return true;
        }
        return future.join().isExpired();
    }

    @Override
    public void put(ID id, T value) {
        Objects.requireNonNull(id, "id must not be null");
        Objects.requireNonNull(value, "value must not be null");
        store.put(id, CompletableFuture.completedFuture(
                new Entry<>(value, ttl == null ? null : Instant.now().plus(ttl))));
        maybeEvictOverCapacity();
    }

    @Override
    public void invalidate(ID id) {
        store.remove(Objects.requireNonNull(id));
    }

    @Override
    public void invalidateAll() {
        store.clear();
    }

    @Override
    public void evictExpired() {
        store.entrySet().removeIf(e -> isFailedOrExpired(e.getValue()));
    }

    /**
     * Approximate-LRU eviction once {@code maxSize} is exceeded: draws a constant
     * sample ({@value #EVICTION_SAMPLE_SIZE} entries) instead of scanning the whole
     * cache, and evicts the least-recently-used entry within it. O(k), independent
     * of the cache size n — keeping the insert path constant-time even at 10^6+ entries.
     */
    private void maybeEvictOverCapacity() {
        if (maxSize <= 0 || store.size() <= maxSize) {
            return;
        }
        ID oldestKey = null;
        long oldestAccess = Long.MAX_VALUE;
        int sampled = 0;

        for (Map.Entry<ID, CompletableFuture<Entry<T>>> e : store.entrySet()) {
            if (sampled++ >= EVICTION_SAMPLE_SIZE) {
                break;
            }
            CompletableFuture<Entry<T>> f = e.getValue();
            if (f.isDone() && !f.isCompletedExceptionally()) {
                long access = f.join().lastAccessNanos.get();
                if (access < oldestAccess) {
                    oldestAccess = access;
                    oldestKey = e.getKey();
                }
            }
        }
        if (oldestKey != null) {
            store.remove(oldestKey);
        }
    }

    @Override
    public int size() {
        return store.size();
    }

    @Override
    public Map<ID, T> snapshot() {
        Map<ID, T> result = new ConcurrentHashMap<>();
        store.forEach((id, future) -> {
            if (future.isDone() && !future.isCompletedExceptionally()) {
                Entry<T> entry = future.join();
                if (!entry.isExpired()) {
                    result.put(id, entry.value);
                }
            }
        });
        return result;
    }
}

File: python/database-driver-plugin/src/database_driver/plugin/utility/cache/default_cache.py
"""Mirror of ``de.lino.database.utility.cache.DefaultCache``."""

from __future__ import annotations

import asyncio
import concurrent.futures
import threading
import time
from collections.abc import Awaitable, Callable
from datetime import UTC, datetime, timedelta
from typing import Any, Generic, TypeVar

from database_driver.api.utils.cache.cache import Cache

ID = TypeVar("ID")
T = TypeVar("T")

_EVICTION_SAMPLE_SIZE = 8


class _Entry(Generic[T]):
    """One cached slot: a thread-safe ``concurrent.futures.Future`` (shared by every
    caller that piggybacked on the same load - the stampede protection), plus the TTL
    and approximate-LRU bookkeeping the eviction paths read."""

    __slots__ = ("expires_at", "future", "last_access_nanos")

    def __init__(self, future: concurrent.futures.Future, expires_at: datetime | None) -> None:
        self.future = future
        self.expires_at = expires_at
        self.last_access_nanos = time.monotonic_ns()

    def is_expired(self) -> bool:
        return self.expires_at is not None and datetime.now(UTC) > self.expires_at

    def is_failed_or_expired(self) -> bool:
        if not self.future.done():
            return False  # still loading - piggyback instead of restarting
        if self.future.exception() is not None:
            return True
        return self.is_expired()

    def touch(self) -> None:
        self.last_access_nanos = time.monotonic_ns()


class DefaultCache(Cache[ID, T]):
    """Default, in-memory implementation of ``Cache``, backed by a lock-guarded dict of
    shared futures - the Python counterpart of the Java edition's
    ``ConcurrentHashMap<ID, CompletableFuture<Entry>>``.

    Complexity: :meth:`get`/:meth:`put`/:meth:`invalidate` are O(1) amortized;
    :meth:`evict_expired` is an O(n) full scan meant for a periodic sweeper, not the hot
    path. Size-based eviction (when ``max_size`` is set) is O(k) with a constant sample
    factor (default 8), independent of n - Redis' "approximated LRU": instead of locating
    the exact oldest entry, a small sample is compared and the oldest candidate within it
    is evicted, keeping the insert path constant-time even with very many entries.

    Design decisions carried over from the Java edition:

    - **Stampede protection:** the install-or-reuse step runs under one lock (the
      analogue of ``ConcurrentHashMap#compute``), so under concurrent requests for the
      same key the loader runs only ONCE - all callers share the same future. The loader
      coroutine itself runs *outside* the lock, exactly as the Java loader runs outside
      the map's bin lock.
    - **Async:** :meth:`get` never blocks the event loop; awaiting a
      ``concurrent.futures.Future`` through ``asyncio.wrap_future`` suspends rather than
      blocks, and works from any event loop - including the private per-call loops the
      section engine's synchronous read path spins up.
    - **Weak-referenceable:** instances deliberately carry no ``__slots__``, since the
      registry's TTL sweeper holds them in a weak-keyed map.
    """

    def __init__(
        self,
        loader: Callable[[ID], Awaitable[T]],
        ttl: timedelta | None = None,
        max_size: int = -1,
    ) -> None:
        """Args:
            loader: Asynchronously supplies an entity when it is not (or no longer)
                cached; must not resolve to ``None``.
            ttl: Validity duration of an entry, ``None`` for unbounded.
            max_size: Maximum number of entries; ``<= 0`` for unbounded. Approximate-LRU
                eviction (O(k)) is triggered once this is exceeded.
        """
        if loader is None:
            raise TypeError("loader must not be None")
        self._loader = loader
        self._ttl = ttl
        self._max_size = max_size
        self._store: dict[ID, _Entry[T]] = {}
        self._lock = threading.Lock()

    async def get(self, id: ID) -> T:
        if id is None:
            raise TypeError("id must not be None")

        load_owner = False
        with self._lock:
            entry = self._store.get(id)
            if entry is None or entry.is_failed_or_expired():
                entry = _Entry(concurrent.futures.Future(), self._expiry())
                self._store[id] = entry
                load_owner = True

        if load_owner:
            try:
                value = await self._loader(id)
                if value is None:
                    raise TypeError("loader must not return None values")
                entry.future.set_result(value)
            except BaseException as failure:
                entry.future.set_exception(failure)

        self._maybe_evict_over_capacity()

        result: T = await asyncio.wrap_future(entry.future)
        entry.touch()  # update access time for approximate LRU
        return result

    def put(self, id: ID, value: T) -> None:
        if id is None:
            raise TypeError("id must not be None")
        if value is None:
            raise TypeError("value must not be None")
        future: concurrent.futures.Future = concurrent.futures.Future()
        future.set_result(value)
        with self._lock:
            self._store[id] = _Entry(future, self._expiry())
        self._maybe_evict_over_capacity()

    def invalidate(self, id: ID) -> None:
        if id is None:
            raise TypeError("id must not be None")
        with self._lock:
            self._store.pop(id, None)

    def invalidate_all(self) -> None:
        with self._lock:
            self._store.clear()

    def evict_expired(self) -> None:
        with self._lock:
            expired = [id for id, entry in self._store.items() if entry.is_failed_or_expired()]
            for id in expired:
                del self._store[id]

    def size(self) -> int:
        return len(self._store)

    def snapshot(self) -> dict[ID, T]:
        result: dict[ID, T] = {}
        with self._lock:
            entries = list(self._store.items())
        for id, entry in entries:
            if entry.future.done() and entry.future.exception() is None and not entry.is_expired():
                result[id] = entry.future.result()
        return result

    def _expiry(self) -> datetime | None:
        return None if self._ttl is None else datetime.now(UTC) + self._ttl

    def _maybe_evict_over_capacity(self) -> None:
        """Approximate-LRU eviction once ``max_size`` is exceeded: draws a constant
        sample instead of scanning the whole cache, and evicts the least-recently-used
        entry within it. O(k), independent of the cache size n."""
        if self._max_size <= 0 or len(self._store) <= self._max_size:
            return

        oldest_key: Any = None
        oldest_access = None

        with self._lock:
            for sampled, (id, entry) in enumerate(self._store.items()):
                if sampled >= _EVICTION_SAMPLE_SIZE:
                    break
                if entry.future.done() and entry.future.exception() is None:
                    if oldest_access is None or entry.last_access_nanos < oldest_access:
                        oldest_access = entry.last_access_nanos
                        oldest_key = id
            if oldest_key is not None:
                self._store.pop(oldest_key, None)
